compliance/virtualmachines/relay/provider/vsock_index_report_provider.go

```diff
@@ -0,0 +1,240 @@
+	defer p.releaseSemaphore()
+
+	defer func(conn net.Conn) {
+		if err := conn.Close(); err != nil {
+			log.Errorf("Failed to close connection: %v", err)
+		}
+	}(conn)
+
+	log.Infof("Handling connection from %s", conn.RemoteAddr())
+
+	indexReport, err := p.receiveAndValidateIndexReport(conn)
+	if err != nil {
+		log.Errorf("Error handling connection from %v: %v", conn.RemoteAddr(), err)
+		return
+	}
+
+	log.Infof("Finished handling connection from %s", conn.RemoteAddr())
+
+	// Send validated report to channel (never blocks - buffer sized to concurrency limit)
+	reportChan <- indexReport
+}
+
+func (p *VsockIndexReportProvider) receiveAndValidateIndexReport(conn net.Conn) (*v1.IndexReport, error) {
+	vsockCID, err := vsock.ExtractVsockCIDFromConnection(conn)
+	if err != nil {
+		return nil, errors.Wrap(err, "extracting vsock CID")
+	}
+
+	data, err := connutil.ReadFromConn(conn, p.maxSizeBytes, p.connectionReadTimeout)
+	if err != nil {
+		return nil, errors.Wrapf(err, "reading from connection (vsock CID: %d)", vsockCID)
+	}
+
+	log.Debugf("Parsing index report (vsock CID: %d)", vsockCID)
+	indexReport, err := parseIndexReport(data)
+	if err != nil {
+		return nil, errors.Wrapf(err, "parsing index report data (vsock CID: %d)", vsockCID)
+	}
+	metrics.IndexReportsReceived.Inc()
+
+	err = validateReportedVsockCID(indexReport, vsockCID)
+	if err != nil {
+		log.Debugf("Error validating reported vsock CID: %v", err)
+		return nil, errors.Wrap(err, "validating reported vsock CID")
+	}
+
+	return indexReport, nil
+}
+
+func parseIndexReport(data []byte) (*v1.IndexReport, error) {
+	report := &v1.IndexReport{}
+
+	if err := proto.Unmarshal(data, report); err != nil {
+		return nil, errors.Wrap(err, "unmarshalling data")
+	}
+	return report, nil
+}
+
+// validateReportedVsockCID ensures the report's vsock CID matches the connection.
+func validateReportedVsockCID(indexReport *v1.IndexReport, connVsockCID uint32) error {
+	if indexReport.GetVsockCid() != strconv.FormatUint(uint64(connVsockCID), 10) {
+		metrics.IndexReportsMismatchingVsockCID.Inc()
+		return errors.Errorf("mismatch between reported (%s) and real (%d) vsock CIDs", indexReport.GetVsockCid(), connVsockCID)
+	}
+	return nil
+}
+
+func (p *VsockIndexReportProvider) stop() {
+	p.stopOnce.Do(func() {
+		log.Info("Stopping connection server")
+		if p.listener != nil {
+			if err := p.listener.Close(); err != nil {
+				log.Errorf("Error closing listener: %v", err)
+			}
+		}
+	})
+}
+
+func (p *VsockIndexReportProvider) acquireSemaphore(parentCtx context.Context) error {
+	semCtx, cancel := context.WithTimeout(parentCtx, p.semaphoreTimeout)
+	defer cancel()
+
+	metrics.SemaphoreQueueSize.Inc()
+	defer metrics.SemaphoreQueueSize.Dec()
+	if err := p.semaphore.Acquire(semCtx, 1); err != nil {
+		reason := "unknown"
+		if errors.Is(err, context.DeadlineExceeded) {
+			log.Debug("Could not acquire semaphore, too many concurrent connections")
+			reason = "concurrency_limit"
```

Comment:
nit
```suggestion
			reason = "reached_concurrency_limit"
```

Reply:
I'd rather leave it as-is to keep the refactor clean (it is now `concurrency_limit` in [master](https://github.com/stackrox/stackrox/blob/e0a5592a60c9bba23204d31871305d95d984f33d/compliance/virtualmachines/relay/server/server.go#L145)). Happy to fix the wording in a follow-up if we prefer that form